src/api/c/random_engine.cpp

```diff
@@ -0,0 +1,357 @@
+
+#include <af/random_engine.h>
+
+#include <af/array.h>
+#include <af/data.h>
+#include <af/defines.h>
+#include <af/dim4.hpp>
+#include <backend.hpp>
+#include <err_common.hpp>
+#include <handle.hpp>
+#include <random_engine.hpp>
+#include <random.hpp>
+#include <MersenneTwister.hpp>
+
+using detail::cfloat;
+using detail::cdouble;
+using detail::uchar;
+using detail::uniformDistribution;
+using detail::normalDistribution;
+using detail::initMersenneState;
+using detail::randu;
+using detail::randn;
+using detail::setSeed;
+using detail::getSeed;
+
+using common::MaxBlocks;
+using common::TableLength;
+using common::MtStateLength;
+using common::pos;
+using common::sh1;
+using common::sh2;
+using common::mask;
+using common::recursion_tbl;
+using common::temper_tbl;
+
+class RandomEngine
+{
+    public :
+    af_random_type type;
+    uintl seed;
+    uintl counter;
+    af_array pos;
+    af_array sh1;
+    af_array sh2;
+    uint mask;
+    af_array recursion_table;
+    af_array temper_table;
+    af_array state;
+
+    RandomEngine(void) : type(AF_RANDOM_THREEFRY), seed(0), counter(0) {}
+};
+
+af_random_engine getRandomEngineHandle(const RandomEngine engine)
+{
+    RandomEngine *engineHandle = new RandomEngine;
+    *engineHandle = engine;
+    return static_cast<af_random_engine>(engineHandle);
+}
+
+af_random_engine DefaultRandomEngine(void)
+{
+    static RandomEngine r;
+    return static_cast<af_random_engine>(&r);
+}
+
+RandomEngine* getRandomEngine(const af_random_engine engineHandle)
+{
+    return (RandomEngine *)engineHandle;
+}
+
+template<typename T>
+static inline af_array uniformDistribution_(const af::dim4 &dims, RandomEngine *e)
+{
+    if (e->type == AF_RANDOM_MERSENNE_GP11213) {
+        return getHandle(uniformDistribution<T>(dims,
+                    getArray<uint>(e->pos),
+                    getArray<uint>(e->sh1),
+                    getArray<uint>(e->sh2),
+                    e->mask,
+                    getArray<uint>(e->recursion_table),
+                    getArray<uint>(e->temper_table),
+                    getArray<uint>(e->state)));
+    } else {
+        return getHandle(uniformDistribution<T>(dims, e->type, e->seed, e->counter));
+    }
+}
+
+template<typename T>
+static inline af_array normalDistribution_(const af::dim4 &dims, RandomEngine *e)
+{
+    if (e->type == AF_RANDOM_MERSENNE_GP11213) {
+        return getHandle(normalDistribution<T>(dims,
+                    getArray<uint>(e->pos),
+                    getArray<uint>(e->sh1),
+                    getArray<uint>(e->sh2),
+                    e->mask,
+                    getArray<uint>(e->recursion_table),
+                    getArray<uint>(e->temper_table),
+                    getArray<uint>(e->state)));
+    } else {
+        return getHandle(normalDistribution<T>(dims, e->type, e->seed, e->counter));
+    }
+}
+
+af_err af_create_random_engine(af_random_engine *engineHandle, af_random_type rtype, uintl seed)
+{
+    try {
```

Comment:
Is any `seed` value accepted?

Also if `seed` is same then return? Or does it require resetting the engine.
Also since `af_random_type` is an enum, whats the fail safe to someone passing garbage arguments?